dotnet/src/SemanticKernel.Abstractions/Orchestration/FunctionResult.cs

```diff
@@ -95,18 +99,12 @@ public FunctionResult(string functionName, ContextVariables variables, object? v
     }
 
     /// <summary>
-    /// Get typed value from metadata.
+    /// Sets function result.
     /// </summary>
-    public bool TryGetVariableValue(string key, out string value)
+    /// <param name="value">The result value.</param>
+    public void SetValue(object? value)
     {
-        if (this.Variables.TryGetValue(key, out string? valueObject))
-        {
-            value = valueObject;
-            return true;
-        }
-
-        value = default!;
-        return false;
+        this.Value = value;
```

Comment:
Why do we need to do this?

I'd expect the `FunctionResult` to be immutable

Reply:
The change is needed to allow the function result to be altered by the invoked hook. Before the changes, the result would be indirectly modified through the context variables within the hook by assigning a new value to the input variable:
<img width="286" alt="image" src="https://github.com/microsoft/semantic-kernel/assets/68852919/189b9c01-2f1e-4493-a0f2-c79b17218160">
After that, the 'CallFunctionInvoked' will create a new instance of 'FunctionResult' and use the input variable from the context as a result.
<img width="367" alt="image" src="https://github.com/microsoft/semantic-kernel/assets/68852919/e0323ac7-1da6-4c0f-b89b-84432bc99b68">

We can consider an alternative design such as adding a new 'UpdateResult(object? newResult)' method to the 'FunctionInvokedEventArgs' class. This would initialize an internal property that would eventually be used when creating a new instance of 'FunctionResult' in the 'CallFunctionInvoked' method.

Reply:
Yeah, let's plan on doing that, think it would be better to restrict when and how the function result can be changed.

Reply:
Yep, I'll fix this shortly after this PR in scope of this issue - https://github.com/microsoft/semantic-kernel/issues/3761